README.md
[![Build Status (GitHub)](https://github.com/FirebirdSQL/firebird/actions/workflows/main.yml/badge.svg?branch=B3_0_Release)](https://github.com/FirebirdSQL/firebird/actions/workflows/main.yml?query=branch%3AB3_0_Release)
[![Build Status (AppVeyor)](https://ci.appveyor.com/api/projects/status/github/FirebirdSQL/firebird?branch=B3_0_Release&svg=true)](https://ci.appveyor.com/project/FirebirdSQL/firebird)

# Firebird README

Firebird is a relational database offering many ANSI SQL standard features that runs on Linux, Windows, MacOS and a variety of Unix platforms. Firebird offers excellent concurrency, high performance, and powerful language support for stored procedures and triggers. It has been used in production systems, under a variety of names, since 1981.

The Firebird Project is a commercially independent project of C/C++ programmers, technical advisors and supporters developing and enhancing a multi-platform relational database management system based on the source code released by Inprise Corporation on 25 July, 2000.

## Basic information

* https://www.firebirdsql.org/
* https://www.firebirdsql.org/en/historical-reference/
* https://www.firebirdsql.org/en/features/

## Documentation

* https://www.firebirdsql.org/en/documentation/
* https://www.firebirdfaq.org/

## Licensing

The source code is released under variants of the Mozilla Public Licence 1.1 (MPL):

* https://www.firebirdsql.org/en/initial-developer-s-public-license-version-1-0/
* https://www.firebirdsql.org/en/interbase-public-license/

## Development

The source code can be found at:  
https://github.com/FirebirdSQL/firebird

Build instructions are here:
https://www.firebirdsql.org/en/building-the-code/

Bugs and feature requests should be submitted at:  
https://github.com/FirebirdSQL/firebird/issues

You may find more details at:  
https://www.firebirdsql.org/en/development/


debian/changelog
firebird3.0 (3.0.11.33703.ds4-2) unstable; urgency=medium

  * Bump policy-compliance version to 4.6.2 (no changes needed)
  * drop dependency on lsb-base

 -- Damyan Ivanov <dmn@debian.org>  Tue, 26 Sep 2023 10:25:06 +0000

firebird3.0 (3.0.11.33703.ds4-1) unstable; urgency=medium

  * New official upstream release

  * add patch adding support for mipsel64 and loongarch
    adapted from upstream commit fcc471e420525dbd4a1dda59ed51487d5ead5921
    (Closes: #1042066, #1051444)
  * refresh patch offsets

 -- Damyan Ivanov <dmn@debian.org>  Mon, 25 Sep 2023 19:40:38 +0000

firebird3.0 (3.0.11.33637.ds4-2) unstable; urgency=medium

  * fix arch-all-only build by overriding only dh_lintian-arch for libib-util
    override mangling.
    Thanks to Ondřej Surý (Closes: #1022579)

 -- Damyan Ivanov <dmn@debian.org>  Mon, 24 Oct 2022 19:36:38 +0000

firebird3.0 (3.0.11.33637.ds4-1) unstable; urgency=medium

  * add gbp.conf for DEP-14 git repository layout
  * switch debian/watch to releases published on firebirdsql.org
    the ones on github aren't always accessible (JavaScript)
  * pack-upstream-snapshot: use mk-origtargz for cleaning upstream sources
    single source of truth about what needs to be removed
  * add files to source repackaging exclusion list to match official releases
    when using git snapshots
  * New upstream snapshot 3.0.11.33637.ds4
    Closes: #1021817 -- undistributable source files
  * drop two files from d/copyright that are not present in the source
  * adapt lintian overrides to version 2.115.3
  * bump Standards-Version from 4.6.0 to 4.6.1 (no changes needed)
  * review debian/copyright, move files under right licensing stanza, update
    copyright years, drop additional files with missing/unclear licenses when
    repackaging

 -- Damyan Ivanov <dmn@debian.org>  Sun, 23 Oct 2022 18:48:29 +0000

firebird3.0 (3.0.10.33601.ds4-1) unstable; urgency=medium

  * New upstream version 3.0.10.33601.ds4
  * update debian/copyright (one new file)
  * rebase no-binary-gbaks.patch to apply to slightly changed upstream

 -- Damyan Ivanov <dmn@debian.org>  Thu, 09 Jun 2022 14:33:54 +0000

firebird3.0 (3.0.9.33560.ds4-3) unstable; urgency=medium

  * add missing trailing slash from debian/firebird-image/ entry of
    debian/clean (Closes: #1010876)
    Thanks to John Paul Adrian Glaubitz

 -- Damyan Ivanov <dmn@debian.org>  Thu, 12 May 2022 09:51:55 +0000

firebird3.0 (3.0.9.33560.ds4-2) unstable; urgency=medium

  * rules: stop overriding -O3 with -O2 during build
  * apply upstream commit adding Documentation to the systemd service
  * apply upstream commit dropping StdErr=syslog from the service

 -- Damyan Ivanov <dmn@debian.org>  Fri, 15 Apr 2022 06:05:22 +0000

firebird3.0 (3.0.9.33560.ds4-1) unstable; urgency=medium

  * New upstream version 3.0.9.33560.ds4
   + refresh/update several patches

 -- Damyan Ivanov <dmn@debian.org>  Mon, 14 Feb 2022 21:09:03 +0000

firebird3.0 (3.0.8.33535.ds4-1) unstable; urgency=medium

  * New upstream version 3.0.8.33535.ds4
    + add several new files to debian/copyright
    + drop deb/kfreebsd-inet6.patch, released upstream
    + refresh several patches
    + adapt out/c++-std.patch
    + adapt out/no-copy-from-icu.patch to the new way upstream borrowed from
      ICU source

 -- Damyan Ivanov <dmn@debian.org>  Sat, 20 Nov 2021 19:23:38 +0000

firebird3.0 (3.0.7.33374.ds4-3) unstable; urgency=medium

  * fix build with missing builds/make.new
    Closes: #997483 -- FTBFS
  * Standards-Version: 4.6.0, no changes necessary

 -- Damyan Ivanov <dmn@debian.org>  Sun, 24 Oct 2021 08:20:02 +0000

firebird3.0 (3.0.7.33374.ds4-2) unstable; urgency=medium

  * work around problems with Turkish locale during configuration
    (Closes: #978074)

 -- Damyan Ivanov <dmn@debian.org>  Fri, 25 Dec 2020 20:12:41 +0000

firebird3.0 (3.0.7.33374.ds4-1) unstable; urgency=medium

  * Use secure URI for Homepage.
    Thanks to the Debian Janitor
  * watch: switch to github, allow any kind of archive suffix
  * New upstream version 3.0.7.33374.ds4
  * refresh two patches

 -- Damyan Ivanov <dmn@debian.org>  Tue, 20 Oct 2020 17:07:16 +0000

firebird3.0 (3.0.6.33328.ds4-2) unstable; urgency=medium

  * Restore charset support broken with -1
    + Revert "remove symlinks in surogate "firebird root" directory"
    + Revert "use --with-fbconf=/etc/firebird/3.0"

 -- Damyan Ivanov <dmn@debian.org>  Wed, 01 Jul 2020 12:01:11 +0000

firebird3.0 (3.0.6.33328.ds4-1) unstable; urgency=medium

  * New upstream version 3.0.6.33328.ds4
  * update copyright summary (two new files)
  * drop core-6343-dupl-pk-after-rollback.patch; released upstream
  * update file offsets in 6 patches
  * convert debian/po/es.po to UTF-8
  * declare compliance with Policy 4.5.0 (no changes needed)
  * recode ods11-index-structure.html to UTF-8
  * fix a typo in license text
  * remove symlinks in surrogate "firebird root" directory and use
    --with-fbconf=/etc/firebird/3.0

 -- Damyan Ivanov <dmn@debian.org>  Mon, 29 Jun 2020 12:19:14 +0000

firebird3.0 (3.0.5.33220.ds4-2) unstable; urgency=medium

  * fix possible duplication of primary key values after an exception and a rollback
    (Closes: #963731)

 -- Damyan Ivanov <dmn@debian.org>  Fri, 26 Jun 2020 08:17:36 +0000

firebird3.0 (3.0.5.33220.ds4-1) unstable; urgency=medium

  * New upstream release
  * remove a couple of files dropped upstream from debian/copyright

 -- Damyan Ivanov <dmn@debian.org>  Wed, 22 Jan 2020 08:04:53 +0000

firebird3.0 (3.0.5.33209.ds4-1) unstable; urgency=medium

  * debian/make_packages.sh: drop 'mkdir' from the list of directories passed
    to mkdir -p. It is harmless, but still wrong
    Thanks to Russell Stuart (Closes: #946663)
  * New upstream snapshot from the B3_0_Release branch
  * copyright:
    + drop two files and add one file
  * debian/copyright_check.pl:
    + adapt to machine-readable debian/copyright
    + skip '.git/' only, not everything beginning with '.git'
  * refresh/rebase patches


src/intl/ld.h
/*
 *	PROGRAM:	Intl
 *	MODULE:		ld.h
 *	DESCRIPTION:	Language Driver data structures
 *
 * The contents of this file are subject to the Interbase Public
 * License Version 1.0 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy
 * of the License at http://www.Inprise.com/IPL.html
 *
 * Software distributed under the License is distributed on an
 * "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * rights and limitations under the License.
 *
 * The Original Code was created by Inprise Corporation
 * and its predecessors. Portions created by Inprise Corporation are
 * Copyright (C) Inprise Corporation.
 *
 * All Rights Reserved.
 * Contributor(s): ______________________________________.
 *
 * 2002.02.15 Sean Leyne - Code Cleanup, removed obsolete ports:
 *                          - MAC ("MAC", "MAC_AUX" and "MAC_CP" defines)
 *                          - XENIX and OS/2
 *
 * 2002.10.27 Sean Leyne - Code Cleanup, removed obsolete "Ultrix" port
 *
 * 2002.10.29 Sean Leyne - Removed obsolete "Netware" port
 *
 * 2002.10.30 Sean Leyne - Removed support for obsolete "MS_DOS" define
 *
 */

/* ------------NOTE NOTE NOTE -----------
 * This file is shared with the Language Driver kit used by PC/Windows
 * products.  This master is in the PC side kit, currently (931018) stored
 * as DBENGINE:/area6/langdrv/h/ld.h
 * with Martin Levy as module owner
 * The file has been ported from ANSI C++ to K&R C, with minimal changes
 * to support re-porting in the future.
 *   C++ comments are surrounded by C comments.
 *   Prototypes are #ifdef'ed out.
 *   Following is a bunch of remappings of InterBase data types to
 *	the type names as used on the PC
 */

#ifndef INTL_LD_H
#define INTL_LD_H

#ifdef CHAR
#undef CHAR
#endif
#define CHAR	SCHAR

#define UINT16	USHORT


/* Following this line is LD.H from Borland Language Driver Kit */


/*
//-----------------------------------------------------------------
//
// LD.H
//
// Language Driver API header file
//
// 08-17-92 MLevy LdLibraryName[ LD_LIBNAME_LEN ] changed to pLdLibraryName.
// 08-05-92 MLevy Exposed LdExit() again.
// 06-09-92 MLevy ICELAND and BorlPORTUGUESE added.
// 05-20-92 MLevy File made.
//
//-----------------------------------------------------------------
*/

//#ifndef __LD_H
//#define __LD_H

/*
//-----------------------------------------------------------------
// DATA TYPE DEFINITIONS
//-----------------------------------------------------------------
*/

#ifndef TRUE
#define           TRUE 1
#endif
#ifndef FALSE
#define           FALSE 0
#endif

/*
//-----------------------------------------------------------------
// API CONSTANT VALUES
//-----------------------------------------------------------------
*/

//#define YES                       TRUE
//#define NO                        FALSE


/*
// LdObject Windows resource type
*/
//#define LDRESOBJECT            1001
//#define LDOBJ_NAME1             101

//#define LDID                   1002

//#define LDIDRESNAME               "LDID"

/*
// Language type (file or resource)
*/
//#define LDOBJRES                  1
//#define LDOBJFILE                 2


//#define SECONDARY_DIFF          256
//#define TERTIARY_DIFF           512
//#define LENGTH_DIFF            1024

/*
//
// SORT/COLLATION TYPE PROVIDED BY LANGUAGE DRIVER
//
*/
//#define ASCII_SORT                1
//#define CODEPAGE_SORT             ASCII_SORT
//#define DICT_SORT                 2
//#define PHONETIC_SORT             3
//#define NAME_SORT                 4

/*
//
// CODE PAGE/CHARACTER SET PLATFORM TYPES
//
*/
//#define DOS_CP                    1
//#define OEM_CP                    DOS_CP
//#define WIN_CP                    2
//#define ANSI_CP                   WIN_CP

/*
// UNIX etc.
*/
//#define SUNOS_CP                  4
//#define VMS_CP                    5
//#define HPUX_CP                   6
//#define AIX_CP                    8
//#define AUX_CP                    9

/*
// SHIFT-JIS (JAPANESE DBCS)
*/
//#define SJIS_CP                 255

/*
// length of descriptive language driver name (incl zero)
*/
//#define LD_SYMNAME_LEN           20

/*
// max length of language driver signature & file name (incl zero)
*/
//#define LD_SFNAME_LEN            20

/*
// max length of language driver resource file name (incl zero)
*/
//#define LD_LIBNAME_LEN           13

//#define LDSUCCESS                 0
//#define LDFAILURE                -1

#define REVERSE					  0x01
#define EXPAND                    0x02
#define SECONDARY                 0x04
#define LOCAL_EXPAND              0x08
#define LOCAL_SECONDARY           0x10

/*
//
// TABLE INDEX CONSTANTS
//
*/

//#define CHARDEF_TBL               0
//#define UPPERCASE_TBL             1
//#define LOWERCASE_TBL             2
//#define EXPAND_TBL                3
//#define COMPRESS_TBL              4
//#define CASESORT_TBL              5
//#define NOCASESORT_TBL            6
//#define SOUNDEX_TBL               7
//#define LICSCP_TBL                8
//#define CPLICS_TBL                9
//#define PRIMALT_TBL              10
//#define ALTPRIM_TBL              11
//#define BASE_TBL                 12
//#define EOF_VAL                  13

//#define NUMBER_TBLS              EOF_VAL + 1

/*
//-----------------------------------------------------------------
*/

struct SortOrderTblEntry {

	UINT16 Primary:8;
	UINT16 Secondary:4;
	UINT16 Tertiary:2;
	UINT16 IsExpand:1;
	UINT16 IsCompress:1;

};

/*
//
// CHARACTER EXPANSION STRUCTURE
//
*/

struct ExpandChar {

	BYTE Ch;
	BYTE ExpCh1;
	BYTE ExpCh2;
};

/*
//
// CHARACTER PAIR TO SINGLE SORT WEIGHT STRUCTURE
//
*/

struct CompressPair {

	BYTE CharPair[2];
	SortOrderTblEntry CaseWeight;
	SortOrderTblEntry NoCaseWeight;
};


/*
//-----------------------------------------------------------------
*/

/*
//-----------------------------------------------------------------
// CHARACTER/SYMBOL DEFINITION VALUES
//-----------------------------------------------------------------
*/

/*
// NO SPECIAL QUALITIES
*/
//#define CH_UNDEFINED           0x00

/*
// A NUMERIC CHARACTER
*/
//#define CH_DIGIT               0x01

/*
// AN UPPERCASE CHARACTER
*/
//#define CH_UPPER               0x02

/*
// A LOWERCASE CHARACTER
*/
//#define CH_LOWER               0x04

/*
// ALPHA CHARACTER
*/
//#define CH_ALPHA               CH_LOWER + CH_UPPER

/*
// ALPHANUMERIC CHARACTER
*/
//#define CH_ALPHANUM            CH_LOWER + CH_UPPER + CH_DIGIT

/*
//-----------------------------------------------------------------
// ID CODES
//-----------------------------------------------------------------
//
// PRODUCT ID
//
*/
//#define PDOX                    0x1
//#define DBASE                   0x2
//#define IBASE                   0x4
//#define OTHER                   0x8
//#define BORLAND                0x10

/*
//
// COUNTRY/LANGUAGE ID
//
*/

/*
//#define US                        1
//#define CANADA                    2
//#define LATINAMER                 3
//#define NEDERLANDS               31
//#define BELGIUM                  32
//#define FRANCE                   33
//#define SPAIN                    34
//#define HUNGARY                  36
//#define YUGOSLAVIA               38
//#define ITALY                    39
//#define SWITZ                    41
//#define CZECH                    42
//#define UK                       44
//#define DENMARK                  45
//#define SWEDEN                   46
//#define NORWAY                   47
//#define POLAND                   48
//#define GERMANY                  49
//#define BRAZIL                   55
//#define INTL                     61
//#define PORTUGAL                351
//#define FINLAND                 358
//#define JAPAN                    81
//#define ICELAND                 354

//#define NORDAN                    DENMARK + NORWAY
//#define SWEDFIN                   SWEDEN  + FINLAND
*/

/*
//
// UNIQUE LANGUAGE DRIVER ID
//
*/

/*
// Paradox
*/
/*
//#define pxUS                      1
//#define pxINTL                    2
//#define pxJAPANESE                3
//#define pxNORDAN                  4
//#define pxNORDAN4                 5
//#define pxSWEDFIN                 6
*/
/*
// dBASE
*/
/*
//#define dbARABIC                  7
//#define dbDANISH                  8
//#define dbDUTCH                   9
//#define dbDUTCH2                 10
//#define dbFINNISH                11
//#define dbFINNISH2               12
//#define dbFRENCH                 13
//#define dbFRENCH2                14
//#define dbGERMAN                 15
//#define dbGERMAN2                16
//#define dbITALIAN                17
//#define dbITALIAN2               18
//#define dbJAPANESE               19
//#define dbSPANISH2               20
//#define dbSWEDISH                21
//#define dbSWEDISH2               22
//#define dbNORWEGIAN              23
//#define dbSPANISH                24
//#define dbUK                     25
//#define dbUK2                    26
//#define dbUS                     27
//#define dbFRENCHCAN              28
//#define dbFRENCHCAN2             29
//#define dbFRENCHCAN3             30
//#define dbCZECH                  31
//#define dbCZECH2                 32
//#define dbGREEK                  33
//#define dbHUNGARIAN              34
//#define dbPOLISH                 35
//#define dbPORTUGUESE             36
//#define dbPORTUGUESE2            37
//#define dbRUSSIAN                38
*/
/*
// Borland
*/
/*
//#define BorlDANISH               39
//#define BorlDUTCH                40
//#define BorlFINNISH              41
//#define BorlFRENCH               42
//#define BorlCANADIAN             43
//#define BorlGERMAN               44
//#define BorlICELANDIC            45
//#define BorlITALIAN              46
//#define BorlJAPANESE             47
//#define BorlNORWEGIAN            48
//#define BorlSPANISH              49
//#define BorlSPANISH2             50
//#define BorlSWEDISH              51
//#define BorlUK                   52
//#define BorlUS                   53
//#define BorlPORTUGUESE           54
//#define dbUS2                    55
*/
/*
// User Defined 201 - 254
*/
/*
//#define USERDEFINED_MIN         201
//#define USERDEFINED_MAX         255
*/

/*
//-----------------------------------------------------------------
// ERROR CODES
//-----------------------------------------------------------------
*/

//#define ERRBASE_LDAPI             0x7000

/*
// no errors at all
*/
//#define LDERR_NONE                LDSUCCESS

/*
// LD memory could be allocated
*/
//#define LDERR_NOMEMORY            ERRBASE_LDAPI + 0xA1

/*
// LD file corrupt/could not be read
*/
//#define LDERR_FILECORRUPT         ERRBASE_LDAPI + 0xA2

/*
// LD version mismatch
*/
//#define LDERR_OLDVERSION          ERRBASE_LDAPI + 0xA3

/*
// LD table not supported
*/
//#define LDERR_NOTSUPPORTED        ERRBASE_LDAPI + 0xA4

/*
// LD file not accessible
*/
//#define LDERR_FILEOPEN            ERRBASE_LDAPI + 0xA5

/*
// LD resource not accessible
*/
//#define LDERR_RESOPEN             ERRBASE_LDAPI + 0xA6

/*
// unknown
*/
//#define LDERR_UNKNOWN             ERRBASE_LDAPI + 0xA7

//#endif


#endif /* INTL_LD_H */
